pkg/openevec/config.go

```diff
@@ -0,0 +1,404 @@
+	Disks             []string `cobrafield:"disks"`
+	Registry          string   `cobrafield:"registry"`
+	OpenStackMetadata bool     `cobrafield:"openstack-metadata"`
+	Profiles          []string `cobrafield:"profile"`
+	DatastoreOverride string   `cobrafield:"datastoreOverride"`
+	AppAdapters       []string `cobrafield:"adapters"`
+	AclOnlyHost       bool     `cobrafield:"only-host"`
+}
+
+type PacketConfig struct {
+	Key string `mapstructure:"key"`
+}
+
+type GcpConfig struct {
+	Key string `mapstructure:"key"`
+}
+
+type EdenSetupArgs struct {
+	Eden     EdenConfig     `mapstructure:"eden"`
+	Adam     AdamConfig     `mapstructure:"adam"`
+	Eve      EveConfig      `mapstructure:"eve"`
+	Redis    RedisConfig    `mapstructure:"redis"`
+	Registry RegistryConfig `mapstructure:"registry"`
+	Packet   PacketConfig   `mapstructure:"packet"`
+	Gcp      GcpConfig      `mapstructure:"gcp"`
+
+	ConfigFile string
+	Runtime    RuntimeConfig
+	ConfigName string
+}
+
+func Merge(dst, src reflect.Value, flags *pflag.FlagSet) {
+	for i := 0; i < dst.NumField(); i++ {
+		if dst.Field(i).Kind() == reflect.Struct {
+			Merge(dst.Field(i), src.Field(i), flags)
+		}
+		if dst.Type().Field(i).Tag != "" {
+			tag := dst.Type().Field(i).Tag.Get("cobraflag")
+			if tag != "" && flags.Changed(tag) {
+				if dst.CanSet() {
+					dst.Set(src)
+				} else {
+					dst = src
+				}
+			}
+		}
+	}
+}
+
+func FromViper(configName, verbosity string) (*EdenSetupArgs, error) {
+	var err error
+	cfg := &EdenSetupArgs{}
+	configNameEnv := os.Getenv(defaults.DefaultConfigEnv)
+	if configNameEnv != "" {
+		configName = configNameEnv
+	}
+	cfg.ConfigFile = utils.GetConfig(configName)
+
+	if verbosity == "debug" {
+		fmt.Println("configName: ", configName)
+		fmt.Println("configFile: ", cfg.ConfigFile)
+	}
+
+	cfg, err = LoadConfig(cfg.ConfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := setUpLogs(os.Stdout, verbosity); err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+func setUpLogs(out io.Writer, level string) error {
+	log.SetOutput(out)
+	lvl, err := log.ParseLevel(level)
+	if err != nil {
+		return err
+	}
+	log.SetLevel(lvl)
+	return nil
+}
+
+func LoadConfig(configFile string) (*EdenSetupArgs, error) {
+	viperLoaded, err := utils.LoadConfigFile(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("error reading config: %s", err.Error())
+	}
+
+	if !viperLoaded {
+		return nil, fmt.Errorf("Viper cannot be loaded")
+	}
+	viper.SetDefault("eve.uefi-tag", defaults.DefaultEVETag)
+
+	cfg := &EdenSetupArgs{}
+
+	if err = viper.Unmarshal(cfg); err != nil {
+		return nil, fmt.Errorf("unable to decode into config struct, %v", err)
+	}
+
+	resolvePath(reflect.ValueOf(cfg).Elem())
+
+	if configFile == "" {
+		configFile, _ = utils.DefaultConfigPath()
+	}
+
+	configName := path.Base(configFile)
+	if pos := strings.LastIndexByte(configName, '.'); pos != -1 {
+		configName = configName[:pos]
+	}
+
+	configCheck(configName)
```

Comment:
We should call the `configCheck` function only from `SetupEden`. The main reason of such check is to not allow re-run setup after config change. We do not support any of upgrade functionality so it is better to have explicit error instead of unpredictable behavior.